File: mobilefoundation/ios/Classes/WLClientMethodHandler.swift
import Flutter
import IBMMobileFirstPlatformFoundation
import UIKit

/// This is a MethodCallManager class for WLClient method calls
public class WLClientMethodHandler {
    static let shared = WLClientMethodHandler()
    static var eventSink: FlutterEventSink?
    var challengeHandlers: [String: MFSecurityCheckChallengeHandler] = [String: MFSecurityCheckChallengeHandler]()
    
    private init()
    {
        // Set up API instance
    }
    
    // This method is used to invoke the challenge handler methods in dart Layer
    // and gets called when the native methods of challenge handler is triggered.
    static func sendEvent(eventName: String, mapObject: [AnyHashable: Any]) {
        if (eventSink != nil) {
            //Send challenge handler method name and it's response data to dart layer.
            let event = NSMutableDictionary()
            event["challengeHandlerMethod"] = eventName
            event["mapObject"] = mapObject
            DispatchQueue.main.async {
                eventSink!(event)
            }
        }
    }
    
    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        do {
            let isValid: Bool = try MFBridgeValidator.sharedInstance().validateMethodCall(call)
            if !isValid {
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
                return
            }
        } catch {
            print("Error \(error)")
            result(FlutterError(code: MFConstants.ERR_METHOD_NAME_CODE, message: error.localizedDescription, details: nil))
            return
        }
        
        switch call.method {
        case Methods.ADDGLOBALHEADER:
            if let args = call.arguments as? [String: Any],
                let headerName = args[Arguments.HEADER_NAME] as? String,
                let headerValue = args[Arguments.HEADER_VALUE] as? String
            {
                WLClient.sharedInstance().addGlobalHeader(headerName, headerValue: headerValue)
                result(nil)
            } else {
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
            }
            break
        case Methods.REMOVEGLOBALHEADER:
            if let args = call.arguments as? [String: Any],
                let headerName = args[Arguments.HEADER_NAME] as? String
            {
                WLClient.sharedInstance()?.removeGlobalHeader(headerName)
                result(nil)
            } else {
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
            }
            break
        case Methods.SETSERVERURL:
            if let args = call.arguments as? [String: Any],
                let serverURL = args[Arguments.SERVER_URL] as? URL
            {
                WLClient.sharedInstance()?.setServerUrl(serverURL)
                result(nil)
            } else {
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
            }
            break
        case Methods.GETSERVERURL:
            let serverURL: String = WLClient.sharedInstance()?.serverUrl()?.absoluteString ?? ""
            result(serverURL)
            break
        case Methods.SETDEVICE_DISPLAYNAME:
            if let args = call.arguments as? [String: Any],
                let displayName = args[Arguments.DEVICE_DISPLAYNAME] as? String
            {
                WLClient.sharedInstance()?.setDeviceDisplayName(displayName, withCompletionHandler:{(error) -> Void in
                    if(error != nil){
                        let mfFailResponse = NSMutableDictionary()
                        mfFailResponse.setValue(String ((error! as NSError).code) , forKey: Arguments.ERROR_CODE)
                        mfFailResponse.setValue((error! as NSError).localizedDescription, forKey: Arguments.ERROR_MSG)
                        result(mfFailResponse)
                        return;
                    }
                    else{
                        result(nil);
                    }
                    
                } )
            } else {
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
            }
            break
        case Methods.GETDEVICE_DISPLAYNAME:
            WLClient.sharedInstance()?.getDeviceDisplayName(completionHandler:{(deviceDisplayName , error) -> Void in
                if(error != nil){
                    let mfFailResponse = NSMutableDictionary()
                    mfFailResponse.setValue(String ((error! as NSError).code) , forKey: Arguments.ERROR_CODE)
                    mfFailResponse.setValue((error! as NSError).localizedDescription, forKey: Arguments.ERROR_MSG)
                    result(mfFailResponse)
                    return;
                }
                else{
                    if(deviceDisplayName == nil){
                        result("")
                        return
                    }
                    result(deviceDisplayName)
                }
            })
            break
        case Methods.SETHEARTBEAT_INTERVAL:
            if let args = call.arguments as? [String: Any],
                let interval = args[Arguments.HEARTBEAT_INTERVAL] as? Int
            {
                WLClient.sharedInstance()?.setHeartBeatInterval(interval);
                result(nil)
                return;
            }
            else {
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
            }
            break
        case Methods.CERTIFICATE_PINNING:
            if let args = call.arguments as? [String: Any],
                let certificates = args[Arguments.CERTIFICATE_FILENAMES] as? Array<String>
            {
                if(certificates.count == 1){
                    WLClient.sharedInstance()?.pinTrustedCertificatePublicKey(fromFile: certificates[0]);
                } else{
                    WLClient.sharedInstance()?.pinTrustedCertificatePublicKey(fromFiles: certificates);
                }
                result(nil)
            } else{
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
            }
            break
        case Methods.WLCLIENT_REGISTER_CHALLENGE_HANDLER:
            if let args = call.arguments as? [String: Any],
                let securityCheck = args[Arguments.SECURITYCHECK_NAME] as? String
            {
                let ch = MFSecurityCheckChallengeHandler(securityCheck: securityCheck)
                WLClient.sharedInstance().registerChallengeHandler(ch)
                challengeHandlers[securityCheck] = ch
                result(nil)
            } else{
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
            }
            break
        case Methods.WLCLIENT_SUBMIT_CHALLENGE_ANSWER:
            if let args = call.arguments as? [String: Any],
                let securityCheck = args[Arguments.SECURITYCHECK_NAME] as? String,
                let challengeAnswer = args[Arguments.ANSWER] as? [AnyHashable : Any]
            {
                let ch = challengeHandlers[securityCheck]
                ch?.submitChallengeAnswer(challengeAnswer)
                result(nil)
            } else{
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
            }
            break
        case Methods.WLCLIENT_CANCEL_CHALLENGE:
            if let args = call.arguments as? [String: Any],
                let securityCheck = args[Arguments.SECURITYCHECK_NAME] as? String
            {
                let ch = challengeHandlers[securityCheck]
                ch?.cancel()
                result(nil)
            } else{
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
            }
            break
        default:
            result(FlutterError(code: MFConstants.ERR_METHOD_NAME_CODE, message: MFConstants.ERR_METHOD_NAME_MESSAGE, details: nil))
        }
    }
    
    private enum Methods{
        public static var ADDGLOBALHEADER: String = "addGlobalHeader"
        public static var REMOVEGLOBALHEADER: String = "removeGlobalHeader"
        public static var SETSERVERURL: String = "setServerUrl"
        public static var GETSERVERURL: String = "getServerUrl"
        public static var SETDEVICE_DISPLAYNAME: String = "setDeviceDisplayName"
        public static var GETDEVICE_DISPLAYNAME: String = "getDeviceDisplayName"
        public static var SETHEARTBEAT_INTERVAL: String = "setHeartbeatInterval"
        public static var CERTIFICATE_PINNING: String = "pinTrustedCertificatesPublicKey"
        public static var WLCLIENT_REGISTER_CHALLENGE_HANDLER: String = "registerChallengeHandler"
        public static var WLCLIENT_SUBMIT_CHALLENGE_ANSWER: String = "submitChallengeAnswer"
        public static var WLCLIENT_CANCEL_CHALLENGE: String = "cancelChallenge"
    }
    
    private enum Arguments{
        public static var HEADER_NAME: String = "headerName"
        public static var HEADER_VALUE = "headerValue"
        public static var SERVER_URL: String = "serverUrl"
        public static var DEVICE_DISPLAYNAME: String = "deviceDisplayName"
        public static var HEARTBEAT_INTERVAL: String = "heartbeatIntervalInSeconds"
        public static var CERTIFICATE_FILENAMES: String = "certificateFileNames"
        public static var SECURITYCHECK_NAME: String = "securityCheckName"
        public static var ANSWER: String = "answer"
        public static var ERROR_CODE: String = "errorCode"
        public static var ERROR_MSG: String = "errorMsg"
    }
    
}

public class ChallengeHandlerStreamHandler: NSObject, FlutterStreamHandler {
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        WLClientMethodHandler.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        WLClientMethodHandler.eventSink = nil
        return nil
    }
}

File: mobilefoundation/android/src/main/java/com/ibm/mobile/mobilefoundation/WLClientMethodHandler.java
package com.ibm.mobile.mobilefoundation;

import android.app.Activity;
import androidx.annotation.NonNull;

import com.worklight.wlclient.WLRequestListener;
import com.worklight.wlclient.api.DeviceDisplayNameListener;
import com.worklight.wlclient.api.WLClient;
import com.worklight.wlclient.api.WLFailResponse;
import com.worklight.wlclient.api.WLResponse;

import org.json.JSONObject;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.BinaryMessenger;


public class WLClientMethodHandler implements MethodChannel.MethodCallHandler {


    private final  WLClient wlClient;
    private static Activity activity;
    private static WLClientMethodHandler instance;
    private static EventChannel.EventSink eventSink;
    private static final String CHALLENGEHANDLER_EVENT = "challengeHandlerEvent";
    Map<String, MFSecurityCheckChallengeHandler> challengeHandlers = new HashMap<String, MFSecurityCheckChallengeHandler>();



    private WLClientMethodHandler() {
        wlClient = WLClient.createInstance(MobilefoundationPlugin.getContext());

        /*Implement a "challengeHandlerEvent" EventChannel to send challenge handler method to dart platform when one of the challenge handler methods of MFSecurityCheckChallengeHandler class is called.
         *The corresponding "challengeHandlerEvent" EventChannel will also be implemented in dart platform for handling received events.
         */
        new EventChannel(MobilefoundationPlugin.getBinaryMessenger(), CHALLENGEHANDLER_EVENT)
                .setStreamHandler(
                        new EventChannel.StreamHandler() {
                            @Override
                            public void onListen(Object arguments, EventChannel.EventSink sink) {
                                eventSink = sink;
                            }

                            @Override
                            public void onCancel(Object arguments) {
                                eventSink = null;
                            }
                        });
    }

    /*This method is used to send challenge handler method to dart layer using "challengeHandlerEvent" EventChannel in native bridge.
     *This method is called when one of the challenge handler method of MFSecurityCheckChallengeHandler.java class is called to send called challenger handler method to dart platform.
     *The corresponding "challengeHandlerEvent" EventChannel in dart platform receives this data to invoke corresponding challenge handler method in dart.
     */
    static void sendEvent(String eventName, Map mapObject) {
        activity = MobilefoundationPlugin.getActivity();
        if (eventSink == null) {
            return;
        }
        final Map<Object, Object> event = new HashMap<>();
        //Send challenge handler method name and it's response data to dart layer.
        event.put("challengeHandlerMethod", eventName);
        event.put("mapObject", mapObject);
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                eventSink.success(event);
            }
        });
    }

    public static WLClientMethodHandler getInstance() {
        if (instance == null) {
            instance = new WLClientMethodHandler();
        }
        return instance;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull final MethodChannel.Result result) {
        try {
            boolean isValid = MFBridgeValidator.getInstance().validateMethodCall(call);
            if (!isValid) {
                result.error(MFConstants.ERR_INCORRECT_ARGS_CODE, MFConstants.ERR_INCORRECT_ARGS_MESSAGE, call.method);
                return;
            }
        } catch (MFBridgeValidator.MethodNotImplementedException ex) {
            result.error(MFConstants.ERR_METHOD_NAME_CODE, MFConstants.ERR_METHOD_NAME_MESSAGE, call.method);
            return;
        }

        if (activity == null) {
            activity = MobilefoundationPlugin.getActivity();
        }

        switch (call.method) {
            case Methods.ADD_GLOBAL_HEADER:
                wlClient.addGlobalHeader((String) call.argument(Arguments.HEADER_NAME),
                        (String) call.argument(Arguments.HEADER_VALUE));
                result.success(null);
                break;
            case Methods.REMOVE_GLOBAL_HEADER:
                wlClient.removeGlobalHeader((String) call.argument(Arguments.HEADER_NAME));
                result.success(null);
                break;

            case Methods.SET_SERVER_URL:
                try {
                    wlClient.setServerUrl(new URL((String) call.argument(Arguments.SERVER_URL)));
                    result.success(null);
                } catch (MalformedURLException e) {
                    result.error(Arguments.SERVER_URL, "Exception setting server URL: " + e.getMessage(), null);
                }
                break;

            case Methods.GET_SERVER_URL:
                result.success(wlClient.getServerUrl().toString());
                break;

            case Methods.WLCLIENT_SETDEVICE_DISPLAYNAME:
                setDeviceDisplayName((String) call.argument(Arguments.DEVICE_DISPLAYNAME), result);
                break;

            case Methods.WLCLIENT_GETDEVICE_DISPLAYNAME:
                getDeviceDisplayName(result);
                break;

            case Methods.WLCLIENT_SETHEARTBEAT_INTERVAL:
                wlClient.setHeartBeatInterval((int) call.argument(Arguments.HEARTBEAT_INTERVAL));
                result.success(null);
                break;

            case Methods.WLCLIENT_CERTIFICATE_PINNING:
                pinTrustedCertificatePublicKey((List<String>) call.argument(Arguments.CERTIFICATE_FILENAMES), result);
                break;

            case Methods.WLCLIENT_REGISTER_CHALLENGEHANDLER:
                MFSecurityCheckChallengeHandler ch = new MFSecurityCheckChallengeHandler((String) call.argument(Arguments.SECURITYCHECK_NAME));
                wlClient.registerChallengeHandler(ch);
                challengeHandlers.put((String) call.argument(Arguments.SECURITYCHECK_NAME), ch);
                result.success(null);
                break;

            case Methods.WLCLIENT_SUBMIT_CHALLENGEANSWER:
                MFSecurityCheckChallengeHandler submitCh = challengeHandlers.get((String) call.argument(Arguments.SECURITYCHECK_NAME));
                submitCh.submitChallengeAnswer(new JSONObject(((HashMap<String, String>) call.argument(Arguments.ANSWER))));
                result.success(null);
                break;

            case Methods.WLCLIENT_CANCEL_CHALLENGE:
                MFSecurityCheckChallengeHandler cancelCh = challengeHandlers.get((String) call.argument(Arguments.SECURITYCHECK_NAME));
                cancelCh.cancel();
                result.success(null);
                break;

            default:
                result.error(MFConstants.ERR_METHOD_NAME_CODE, MFConstants.ERR_METHOD_NAME_MESSAGE, null);

        }
    }

    private void setDeviceDisplayName(String deviceDisplayName, final MethodChannel.Result result) {
        wlClient.setDeviceDisplayName(deviceDisplayName, new WLRequestListener() {
            @Override
            public void onSuccess(final WLResponse wlResponse) {
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        result.success(wlResponse.getResponseText());
                    }
                });
            }

            @Override
            public void onFailure(final WLFailResponse wlFailResponse) {
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        result.error(wlFailResponse.getErrorStatusCode(), wlFailResponse.getErrorMsg(), wlFailResponse.toString());
                    }
                });
            }
        });
    }

    private void getDeviceDisplayName(final MethodChannel.Result result) {
        wlClient.getDeviceDisplayName(new DeviceDisplayNameListener() {
            @Override
            public void onSuccess(final String deviceDisplayName) {
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        result.success(deviceDisplayName);
                    }
                });

            }

            @Override
            public void onFailure(final WLFailResponse wlFailResponse) {
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        result.error(wlFailResponse.getErrorStatusCode(), wlFailResponse.getErrorMsg(), wlFailResponse.toString());
                    }
                });
            }
        });
    }

    private void pinTrustedCertificatePublicKey(List<String> certificateFileNames, MethodChannel.Result result) {
        try {
            if (null != certificateFileNames) {
                if (certificateFileNames.size() == 1) {
                    wlClient.pinTrustedCertificatePublicKey(certificateFileNames.get(0));
                } else {
                    String certificateNames[] = new String[certificateFileNames.size()];
                    for (int j = 0; j < certificateFileNames.size(); j++) {
                        certificateNames[j] = certificateFileNames.get(j);
                    }

                    wlClient.pinTrustedCertificatePublicKey(certificateNames);
                    result.success(null);
                }
            }
        } catch (Exception e) {
            result.error("pinTrustedCertificatesPublicKey", "Certificate pinning error :" + e.getMessage(), null);
        }
    }

    private static class Methods {
        public static final String ADD_GLOBAL_HEADER = "addGlobalHeader";
        public static final String REMOVE_GLOBAL_HEADER = "removeGlobalHeader";
        public static final String SET_SERVER_URL = "setServerUrl";
        public static final String GET_SERVER_URL = "getServerUrl";
        public static final String WLCLIENT_SETDEVICE_DISPLAYNAME = "setDeviceDisplayName";
        public static final String WLCLIENT_GETDEVICE_DISPLAYNAME = "getDeviceDisplayName";
        public static final String WLCLIENT_SETHEARTBEAT_INTERVAL = "setHeartbeatInterval";
        public static final String WLCLIENT_CERTIFICATE_PINNING = "pinTrustedCertificatesPublicKey";
        public static final String WLCLIENT_REGISTER_CHALLENGEHANDLER = "registerChallengeHandler";
        public static final String WLCLIENT_SUBMIT_CHALLENGEANSWER = "submitChallengeAnswer";
        public static final String WLCLIENT_CANCEL_CHALLENGE = "cancelChallenge";
    }

    private static class Arguments {
        public static final String HEADER_NAME = "headerName";
        public static final String HEADER_VALUE = "headerValue";
        public static final String SERVER_URL = "serverUrl";
        public static final String DEVICE_DISPLAYNAME = "deviceDisplayName";
        public static final String HEARTBEAT_INTERVAL = "heartbeatIntervalInSeconds";
        public static final String CERTIFICATE_FILENAMES = "certificateFileNames";
        public static final String SECURITYCHECK_NAME = "securityCheckName";
        public static final String ANSWER = "answer";
    }
}
